### ISSUE TITLE
Top-p halves the generation speed in the Llama example

### ISSUE THREAD

**Issue by Ayuei** [huggingface/candle#2147]
Hi, I'm reaching out about a particular problem I'm running into. Using top-p with the logits sampler seems to halve the generation speed, and I'm not sure why.

```
❯ ./target/release/examples/llama --prompt "Jack and Jill" --top-p 0.9
loading the model weights from meta-llama/Meta-Llama-3-8B
starting the inference loop
Jack and Jill is a fun, affordable hotel for families. With 48 rooms including suites and cabins as well as a campground, there's something here for everyone.
This hotel is perfect if you are looking to stay in the country but not too far from town. The hotel has large units with kitchens, wood burning fireplaces, and decks overlooking acres of meadows or the mountain.
Located just off highway 20 between Tahoe City and Truckee, the Cedar House Sport Hotel features luxurious accommodation surrounded by beautiful mountains and streams.

103 tokens generated (34.67718999077969 token/s)
```

```
❯ ./target/release/examples/llama --prompt "Jack and Jill"
loading the model weights from meta-llama/Meta-Llama-3-8B
starting the inference loop
Jack and Jill is a children's nursery rhyme that has been recorded for over two centuries.
The rhyme counts down from ten and is associated with other verses about the names Jack and Jill. Children usually play a game where they hold hands in a circle, and one child claps his/her hands while singing the rhyme.
In this version: Take turns repeating lines. Line 1 is said by each person at the same time until everyone says line 3, then repeat line 4 etc.

96 tokens generated (69.52924165700331 token/s)
```

This correlates with GPU utilization dropping from ~95% to 47% when top-p sampling is enabled.

### With top-p
![image](https://github.com/huggingface/candle/assets/18545131/d7a44160-0caf-4bef-a259-8db4562b810f)

### Without top-p
![image](https://github.com/huggingface/candle/assets/18545131/21e00f33-24ac-4d65-8085-28af758d0bbc)

Tested on a H100 94GB.

Love this library, thanks in advance 😄

**Comment by LaurentMazare** [comment]
The issue there is that the sampling takes place on the cpu at the moment and so we have to compute the multinomial bits for all the vocab size (which is very large for llama 3 at 128k). However when using top-p, we do a first pass that just selects the top-p values in linear time and then only do the multinomial part on the logits that are left (which is a very small number). We're working on putting this part more on the gpu so hopefully this will improve things significantly but at the moment the recommendation is certainly to use top-p :)

**Comment by LaurentMazare** [comment]
Sorry I messed things up between top-k and top-p in the previous comment, I've updated it so that hopefully it makes more sense now, in a nutsheel if you want to use top-p you can combine it with top-k and it's likely to improve performance quite a bit.
I've also added support for top-k in the llama example in #2150  so that it's easier to illustrate it, here are the timings I get on a H100  SXM (80G), using top-k makes for a big improvement but hopefully with the upcoming sampling improvement it should not be needed anymore.
```
$ cargo run --example llama -r --features cuda -- --prompt "Jack and Jill" --top-p 0.9 -n 100
     Running `target/release/examples/llama --prompt 'Jack and Jill' --top-p 0.9 -n 100`
...
100 tokens generated (35.17282300703097 token/s)

$ cargo run --example llama -r --features cuda -- --prompt "Jack and Jill" --top-p 0.9 -n 100 --top-k 200
loading the model weights from meta-llama/Meta-Llama-3-8B
...
100 tokens generated (81.1718374444453 token/s)
```

**Comment by Ayuei** [comment]
That makes sense, the vocabulary size is a major problem here alongside the expensive sorting step. I tested on CPU and metal as well which reflects this.

I had assumed everything was done on GPU and was wondering why the utilization dropped. Looking forward to the upcoming sampling improvement 😃

Looks like top-p + top-k is a good workaround, thanks!
